design/oauth2-device-authorization-grant.md

```diff
@@ -0,0 +1,144 @@
+
+### Which user interaction does keycloak should support?
+
+User Code Verification First is adopted by major IdPs like Google, Microsoft, and Salesforce. So it seems to be a common practice. Also, it can provide a more flexible authentication flow per device client. Keycloak already provides a feature that a client can use specific browser flow by using the Authentication Flow Overrides option. Therefore, it's suitable for supporting User Code Verification First.
+
+## Consent
+
+The spec says that the authorization server SHOULD display information about the device so that the person can notice if a software client was attempting to impersonating a hardware device. So It should always display consent screen when end-user do log in even if consent required is true or the user already consented.
+
+## User Code Format
+
+The spec says about user code format patterns in [6.1. User Code Recommendations](https://tools.ietf.org/html/draft-ietf-oauth-device-flow-15#section-6.1). It's good to be able to provide an SPI for user code format customization. We implement a case-insensitive eight-letter format as a default implementation of the SPI.
+
+## Configuration
+
+### Realm Configuration
+
+In the token tab of the realm configuration, we add two options for the spec.
+
+* **OAuth 2.0 Device Code Lifespan:** Max time before the device code and user code are expired. This value needs to be a long enough lifetime to be usable (allowing the user to retrieve their secondary device, navigate to the verification URI, login, etc.) but should be sufficiently short to limit the usability of a code obtained for phishing. The default value is 600 seconds (10 minutes).
+* **OAuth 2.0 Device Polling Interval:** The minimum amount of time in seconds that the client should wait between polling requests to the token endpoint. The default value is 5 seconds.
+
+### Client Configuration
+
+In the client edit page for openid-connect protocol, we add an option for the spec.
+
+* **OAuth 2.0 Device Authorization Grant Enabled:** This enables support for OAuth 2.0 Device Authorization Grant, which means that the client is an application on the device that has limited input capabilities or lacks a suitable browser.
+
+Also, we add some extra options when the above option is enabled.
+
+* **OAuth 2.0 Device Code Lifespan:** This is used for overriding realm configuration per client.
+* **OAuth 2.0 Device Polling Interval:** This is used for overriding realm configuration per client.
+
+## Database changes
+
+There are no database schema changes since we can store new configuration for the spec into `REALM_ATTRIBUTE` and `CLIENT_ATTRIBUTES` tables.
```

Comment:
I am wondering whether not doing any schema changes and storing all introduced new fields as custom attributes is actually a good idea. I would expect custom attributes are used where Keycloak does not understand the data in any way, it is just for clients to have an additional place to store custom data. If Keycloak needs to understand the data, I think it is better to include it in the domain model and make sure the data is consistent (e.g. type, constraints).

Reply:
I think using attribute is preferred way to avoid DB migration lately.

For example, KEYCLOAK-7688 Offline Session Max for Offline Token was implemented with same way:
http://lists.jboss.org/pipermail/keycloak-dev/2018-June/010944.html

Also, @stianst mentioned about using attributes in [KEYCLOAK-7560 Refactor Token Sign and Verify by Token Signature SPI](https://github.com/keycloak/keycloak/pull/5260#issuecomment-397509607) as follows.

> For token signature algorithm in realm and clients just use attributes. You can add getter/setter methods to the models, but the JPA entities should just use attributes. Having specific fields on the entities is really the old way of doing it, which requires more work adding, but also more work for the migration.